src/BatteryStateOfChargeService.cpp

```diff
@@ -35,29 +36,49 @@ namespace
 BatteryStateOfChargeService::BatteryStateOfChargeService(double initialStateOfChargePercent)
 : initialStateOfChargePercent_(initialStateOfChargePercent)
 {
+
 }
 
 BatteryStateOfChargeService::~BatteryStateOfChargeService()
 {
+
 }
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return totalcurrent;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    if (newcurrent<=0)
+        qDebug()<<"is charging";
+    else
+        return false;
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    return et;
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
-{
+{   
+    newcurrent = batteryData.current;
+    newtime = batteryData.time;
+    s1 = QTime(0,0,0).secsTo(newtime);
+    s2 = QTime(0,0,0).secsTo(oldtime);
+    timeused = s1-s2;
+    totalcurrent = newcurrent - oldcurrent;
+    double avg = (fabs(newcurrent)+fabs(oldcurrent)) *3600 / timeused;
```

Comment:
The build fails because _fabs_ was not declared in this scope, and I can't find it anywhere else in your program. It looks like you forgot to define it.